server/cellsv2.go

```diff
@@ -0,0 +1,220 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/bouk/httprouter"
+	chronograf "github.com/influxdata/chronograf/v2"
+)
+
+type cellV2Links struct {
+	Self string `json:"self"`
+}
+
+type cellV2Response struct {
+	chronograf.Cell
+	Links cellV2Links `json:"links"`
+}
+
+func (r cellV2Response) MarshalJSON() ([]byte, error) {
+	vis, err := chronograf.MarshalVisualizationJSON(r.Visualization)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(struct {
+		chronograf.CellContents
+		Links         cellV2Links     `json:"links"`
+		Visualization json.RawMessage `json:"visualization"`
+	}{
+		CellContents:  r.CellContents,
+		Links:         r.Links,
+		Visualization: vis,
+	})
+}
+
+func newCellV2Response(c *chronograf.Cell) cellV2Response {
+	return cellV2Response{
+		Links: cellV2Links{
+			Self: fmt.Sprintf("/chronograf/v1/cells/%s", c.ID),
+		},
+		Cell: *c,
+	}
+}
+
+// CellsV2 returns all cells within the store.
+func (s *Service) CellsV2(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	// TODO: support filtering via query params
+	cells, _, err := s.Store.Cells(ctx).FindCells(ctx, chronograf.CellFilter{})
+	if err != nil {
+		Error(w, http.StatusInternalServerError, "Error loading cells", s.Logger)
+		return
+	}
+
+	s.encodeGetCellsResponse(w, cells)
+}
+
+type getCellsLinks struct {
+	Self string `json:"self"`
+}
+
+type getCellsResponse struct {
+	Links getCellsLinks    `json:"links"`
+	Cells []cellV2Response `json:"cells"`
+}
+
+func (s *Service) encodeGetCellsResponse(w http.ResponseWriter, cells []*chronograf.Cell) {
+	res := getCellsResponse{
+		Links: getCellsLinks{
+			Self: "/chronograf/v2/cells",
+		},
+		Cells: make([]cellV2Response, 0, len(cells)),
+	}
+
+	for _, cell := range cells {
+		res.Cells = append(res.Cells, newCellV2Response(cell))
+	}
+
+	encodeJSON(w, http.StatusOK, res, s.Logger)
+}
+
+// NewCellV2 creates a new cell.
+func (s *Service) NewCellV2(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	req, err := decodePostCellRequest(ctx, r)
+	if err != nil {
+		Error(w, http.StatusBadRequest, err.Error(), s.Logger)
+		return
+	}
+	if err := s.Store.Cells(ctx).CreateCell(ctx, req.Cell); err != nil {
+		Error(w, http.StatusInternalServerError, fmt.Sprintf("Error loading cells: %v", err), s.Logger)
+		return
+	}
+
+	s.encodePostCellResponse(w, req.Cell)
+}
+
+type postCellRequest struct {
+	Cell *chronograf.Cell
+}
+
+func decodePostCellRequest(ctx context.Context, r *http.Request) (*postCellRequest, error) {
+	c := &chronograf.Cell{}
+	if err := json.NewDecoder(r.Body).Decode(c); err != nil {
+		return nil, err
+	}
+	return &postCellRequest{
```

Comment:
What do you think of adding a ValidCell function and test?

Reply:
will do